librosa/_typing.py

```diff
@@ -11,6 +12,25 @@
 _SequenceLike = Union[Sequence[_T], np.ndarray]
 _ScalarOrSequence = Union[_T, _SequenceLike[_T]]
 
+_ResampleType = Union[
+    Literal["soxr_vhq"],
+    Literal["soxr_hq"],
+    Literal["soxr_mq"],
+    Literal["soxr_lq"],
+    Literal["soxr_qq"],
+    Literal["kaiser_best"],
+    Literal["kaiser_fast"],
+    Literal["fft"],
+    Literal["scipy"],
+    Literal["polyphase"],
+    Literal["linear"],
+    Literal["zero_order_hold"],
+    Literal["sinc_best"],
+    Literal["sinc_medium"],
+    Literal["sinc_fastest"],
+    str,
```

Comment:
Why have str here?  Seems like this would defeat the purpose of having an explicitly enumerated union of literals.

Reply:
Consider this example user code:

```python
def user_func(res_type: str):
    librosa.resample(..., res_type=res_type)
```

Would you like it to throw a type error? The user would have to add `#type: ignore` or an ugly cast to a Union of Literals. What if the string is read from a config file?

If you think that it's better to go back to `str` for `res_type`, I'm fine with that

Reply:
That's an interesting point, and not something I'd considered.

I suppose it technically should be a type error - if the user can't guarantee that res_type is one of the supported enums, then it's either a type error here or a runtime error later.  Requiring the user to explicitly ignore types in this setting makes sense to me.

On the other hand, I see how that's pretty annoying in practice, and it may be better to just leave it as `str`.  My only point originally was that if we're going to allow `str`, then the remaining Literals are not doing anything for type hinting.  It'd be like pattern matching against `foo|bar|.*`.  Having these literals might be useful as a soft hint to a user of what's supported, but they can get that from the docstring anyway.

Barring any strong reasons for keeping the literals here, I guess I'd lean toward reverting to `str`.

Reply:
You're right that there's no point in the `Literal` if `str` is there, I didn't realize that. So we should choose.

Several observations:
 1. mypy doesn't infer `Literal` for variables, but pyright does (both by design):
```python
def test(x: Literal["x", "y"]):
    pass

x = "x"
test(x) # error in mypy, ok in pyright
```
2. mypy does allow the call when the variable is annotated with Final or Literal:
```python
x2: Final = "x"
test(x2) # no error
x3: Literal["x"] = "x"
test(x3) # no error
```
 3. Some function parameters in stdlib are annotated with Literals:
```python
i = 3
byte_order = "big"
i.to_bytes(1, byte_order) # error in mypy 

import csv
action = "raise"
csv.DictWriter(..., ..., extrasaction=action) # error in mypy

import zipfile
mode = "r"
zipfile.ZipFile(..., mode=mode)  # error in mypy

import tkinter
aaa = "normal"
tkinter.Button(default=aaa) # error in mypy
```

I'm leaning towards the Literals because:
* We can probably assume that the majority of calls use a literal string argument (e.g. `librosa.resample(res_type="fft")`). So we get better type checking for the majority.
* Given the precedents in stdlib, we can assume it's safe for us too. Probably tooling would be improved at some point (e.g. inference in mypy) or better standards would emerge
* For more 'advanced' cases (see below), false positives can be silenced easily

```python
from librosa.beat import beat_track

def my_beat_track(use_frames: bool):
    beat_track(units="frames" if use_frames else "samples") # no error
    units = "frames" if use_frames else "samples"
    beat_track(units=units) # error in mypy
    units2: Literal["frames", "samples"] = "frames" if use_frames else "samples"
    beat_track(units=units2) # no error

def my_beat_track2():
   res_type_str = input("Please enter res_type: ")
   assert res_type_str in ("samples", "frames")
   beat_track(units=res_type_str) # error in mypy
   res_type = cast(Literal["samples", "frames"], res_type_str) # needed for mypy (or use type: ignore)
   beat_track(units=res_type) # no error
```

Reply:
Ok, so let's remove literals when they don't affect return value.

Reply:
Sounds good - thanks for hashing this out!

Reply:
> it seems like most of the rationale for using literals is to support overloading

> Ok, so let's remove literals when they don't affect return type.

That's a bit sad. From my experience, the usage of literals goes beyond overloading and return type narrowing. It simply communicates the allowed values clearly via the function signature itself, enabling verification in the type system and (perhaps most importantly from a practical perspective) auto-completion on the allowed values:

![image](https://user-images.githubusercontent.com/3620703/206664221-7e263a82-20d7-4774-93fe-1607f3adb45a.png)

Why not expose the literal type to the user? Then user code can be written as:

> def user_func(res_type: ResType):
    librosa.resample(..., res_type=res_type)

Or does `res_type` really require a `str` branch because `resample` really can consume arbitrary strings?

> What if the string is read from a config file?

If the user code uses a validation library like pydantic that supports literals, the validation process would make sure that the value is in the allowed list, which makes usage of literals very clean (no ["shotgun parsing"](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/)). If the user code uses e.g. a plain json/yaml parser, the return type will be `Any`, which would also "type check". So that shouldn't be a problem.

Reply:
I agree, it's a tough call, but as we've already spent considerable time on this, I think this can be addressed in a future pr if needed

Reply:
Perhaps in the future behavior of type checkers regarding literals would also be more consistent